hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/ContractStateMigrator.java

```diff
@@ -0,0 +1,417 @@
+     * returns the copy for further writing.
+     */
+    public interface StateFlusher extends UnaryOperator<WritableKVState<SlotKey, SlotValue>> {}
+
+    public static Status migrateFromContractStoreVirtualMap(
+            @NonNull final VirtualMapLike<ContractKey, IterableContractValue> fromState,
+            @NonNull final WritableKVState<SlotKey, SlotValue> initialToState,
+            @NonNull final StateFlusher stateFlusher) {
+        return migrateFromContractStoreVirtualMap(
+                fromState, initialToState, stateFlusher, new ArrayList<>(), false, DEFAULT_THREAD_COUNT);
+    }
+
+    @NonNull
+    public static Status migrateFromContractStoreVirtualMap(
+            @NonNull final VirtualMapLike<ContractKey, IterableContractValue> fromState,
+            @NonNull final WritableKVState<SlotKey, SlotValue> initialToState,
+            @NonNull final StateFlusher stateFlusher,
+            @NonNull final List<String> validationFailures) {
+        return migrateFromContractStoreVirtualMap(
+                fromState, initialToState, stateFlusher, validationFailures, true, DEFAULT_THREAD_COUNT);
+    }
+
+    @NonNull
+    public static Status migrateFromContractStoreVirtualMap(
+            @NonNull final VirtualMapLike<ContractKey, IterableContractValue> fromState,
+            @NonNull final WritableKVState<SlotKey, SlotValue> initialToState,
+            @NonNull final StateFlusher stateFlusher,
+            @NonNull final List<String> validationFailures,
+            final boolean doFullValidation,
+            final int threadCount) {
+        requireNonNull(fromState);
+        requireNonNull(initialToState);
+        requireNonNull(stateFlusher);
+        requireNonNull(validationFailures);
+
+        validationFailures.clear();
+
+        final var migrator = new ContractStateMigrator(fromState, initialToState, stateFlusher)
+                .withValidation(doFullValidation)
+                .withThreadCount(threadCount);
+
+        final var status = migrator.doit(validationFailures);
+        if (status != Status.SUCCESS) {
+            final var formattedFailedValidations =
+                    validationFailures.stream().collect(Collectors.joining("\n   ***", "   ***", "\n"));
+            log.error("{}: transform validations failed:\n{}", LOG_CAPTION, formattedFailedValidations);
+            throw new BrokenTransformationException(LOG_CAPTION + ": transformation didn't complete successfully");
+        }
+        return status;
+    }
+
+    /** A reasonable guess as to the number of threads to devote to the source tree traversal.
+     *
+     *  a) Don't wish to use a configuration option because I want to be able to use this class in test code or a CLI
+     *  tool that doesn't necessarily have the services (or platform) configuration system (easily) available.
+     *  b) And though `availableProcessors()` may in fact give different results during the _same_ JVM invocation: IDC.
```

Comment:
😂

Reply:
(actually I think I meant 2/3 not 3/2 now I look at it again ...)

Reply:
The 😂 is mostly for `IDC` haha

Reply:
(yeah I got that! I figure: reading code tasks the mind ... why not provide the reader a break or two along the way, make it easier?)